lib/grandpa/engine.go

```diff
@@ -0,0 +1,475 @@
+
+const (
+	determinePrevote engineAction = iota
+	determinePrecommit
+	alreadyFinalized
+	finalize
+)
+
+type finalizationEngine struct {
+	grandpaService *Service
+
+	timeoutStop time.Duration
+	stopCh      chan struct{}
+	engineDone  chan struct{}
+	actionCh    chan engineAction
+}
+
+func newFinalizationEngine(service *Service) *finalizationEngine {
+	return &finalizationEngine{
+		grandpaService: service,
+		timeoutStop:    5 * time.Second,
+		actionCh:       make(chan engineAction),
+		stopCh:         make(chan struct{}),
+		engineDone:     make(chan struct{}),
+	}
+}
+
+func (f *finalizationEngine) Stop() (err error) {
+	if f.stopCh == nil {
+		return nil
+	}
+
+	close(f.stopCh)
+	select {
+	case <-f.engineDone:
+	case <-time.After(f.timeoutStop):
+		return fmt.Errorf("%w", errTimeoutWhileStoping)
+	}
+
+	f.stopCh = nil
+	close(f.actionCh)
+	return nil
+}
+
+func (f *finalizationEngine) Start() (err error) {
+	defer func() {
+		close(f.engineDone)
+		fmt.Println("RETURNING FINALIZATION ENGINE")
+	}()
+
+	gossipInterval := f.grandpaService.interval
+	determinePrevoteTimer := time.NewTimer(gossipInterval * 2)
+	determinePrecommitTimer := time.NewTimer(gossipInterval * 4)
+
+	var precommited bool = false
+
+	for !precommited {
+		select {
+		case <-f.stopCh:
+			if !determinePrevoteTimer.Stop() {
+				<-determinePrevoteTimer.C
+			}
+
+			if !determinePrecommitTimer.Stop() {
+				<-determinePrecommitTimer.C
+			}
+
+			return nil
+
+		case <-determinePrevoteTimer.C:
+			alreadyCompletable, err := f.grandpaService.checkRoundAlreadyCompletable()
+			if err != nil {
+				return fmt.Errorf("checking round is completable: %w", err)
+			}
+
+			if alreadyCompletable {
+				f.actionCh <- alreadyFinalized
+				return nil
+			}
+
+			f.actionCh <- determinePrevote
+
+		case <-determinePrecommitTimer.C:
+			alreadyCompletable, err := f.grandpaService.checkRoundAlreadyCompletable()
+			if err != nil {
+				return fmt.Errorf("checking round is completable: %w", err)
+			}
+
+			if alreadyCompletable {
+				f.actionCh <- alreadyFinalized
+				return nil
+			}
+
+			prevoteGrandpaGhost, err := f.grandpaService.getPreVotedBlock()
+			if err != nil {
+				return fmt.Errorf("getting grandpa ghost: %w", err)
+			}
+
+			total, err := f.grandpaService.getTotalVotesForBlock(prevoteGrandpaGhost.Hash, prevote)
+			if err != nil {
+				return fmt.Errorf("getting grandpa ghost: %w", err)
+			}
+
+			if total <= f.grandpaService.state.threshold() {
+				determinePrecommitTimer.Reset(gossipInterval * 4)
+				break
+			}
+
+			latestFinalizedHash := f.grandpaService.head.Hash()
+			isDescendant, err := f.grandpaService.blockState.IsDescendantOf(
+				latestFinalizedHash, prevoteGrandpaGhost.Hash)
+			if err != nil {
+				return fmt.Errorf("checking grandpa ghost ancestry: %w", err)
+			}
+
+			if !isDescendant {
+				panic("block with supermajority does not belong to the latest finalized block chain")
+			}
+
+			f.actionCh <- determinePrecommit
+			precommited = true
```

Comment:
I kinda like using the loop labels here and breaking out of this explicitly.  It's easier to read, since it's not obvious this breaks out of the loop on line 377.

Reply:
I have a different point of view. When I call break inside a select it jump out the select block, and in this case is the same as going to the next loop iteration, and there is only one `break` in this entire select block, I don't see this code as convoluted/not obvious

Reply:
I mean using a loop label and explicit breaking.
```
main:
	for {
		select {
		case <-f.stopCh:
			...
		case <-determinePrevoteTimer.C:
			...
		case <-determinePrecommitTimer.C:
			...
			break main
		}
	}
```

Reply:
The line 441 is `precommited = true` which means that you're trying to break out of the loop in on line 377.  Am I misunderstanding something?  I was just suggesting adding a loop label so you can use the `break <loop name>` explicitly.

Reply:
Oh, right, I set `precommited = true` then I go out the for loop. I choose this way using the variable because it looks very explicit and given the fact that Go enables loop conditionals, your way achieves the same result but uses labels which I don't think is a good fit for this case since Go enable this loop conditional